source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -982,20 +1030,30 @@ void Filter::sendBodyChunk(ProcessorState& state, ProcessorState::CallbackState
 
 void Filter::sendTrailers(ProcessorState& state, const Http::HeaderMap& trailers,
                           bool observability_mode) {
+  // Skip if the trailers is already sent to the server.
+  if (state.trailersSentToServer()) {
+    return;
+  }
+  state.setTrailersSentToServer(true);
+
   ProcessingRequest req;
   req.set_observability_mode(observability_mode);
   addAttributes(state, req);
   addDynamicMetadata(state, req);
   auto* trailers_req = state.mutableTrailers(req);
   MutationUtils::headersToProto(trailers, config_->allowedHeaders(), config_->disallowedHeaders(),
                                 *trailers_req->mutable_trailers());
-
   if (observability_mode) {
     ENVOY_LOG(debug, "Sending trailers message in observability mode");
   } else {
-    state.onStartProcessorCall(std::bind(&Filter::onMessageTimeout, this),
-                               config_->messageTimeout(),
-                               ProcessorState::CallbackState::TrailersCallback);
+    if (state.callbackState() == ProcessorState::CallbackState::Idle) {
+      state.onStartProcessorCall(std::bind(&Filter::onMessageTimeout, this),
+                                 config_->messageTimeout(),
+                                 ProcessorState::CallbackState::TrailersCallback);
+    } else {
+      state.onStartProcessorCall(std::bind(&Filter::onMessageTimeout, this),
+                                 config_->messageTimeout(), state.callbackState());
+    }
```

Comment:
would this be easier to read if you wrote it as:
```
ProcessorState::CallbackState state = state.callbackState();
if (state == ProcessorState::CallbackState::Idle) {
  state = ProcessorState::CallbackState::TrailersCallback;
}
state.onStartProcessorCall(std::bind(&Filter::onMessageTimeout, this), config_->messageTimeout(), state);
```

It avoids repeating most of a complex call-site, and removes an extra call to state.callbackState(), but more importantly it shows the transition from 'idle' to 'trailers' more explicitly, whereas in other cases 'state' stays as it was.

Reply:
done